designer/server/src/main/scala/pl/touk/nussknacker/ui/validation/CustomActionValidator.scala

```diff
@@ -0,0 +1,129 @@
+import pl.touk.nussknacker.engine.api.definition.MandatoryParameterValidator
+import pl.touk.nussknacker.engine.api.deployment.CustomActionCommand
+import pl.touk.nussknacker.engine.api.parameter.ParameterName
+import pl.touk.nussknacker.engine.deployment.{CustomActionDefinition, CustomActionParameter}
+import pl.touk.nussknacker.engine.graph.expression.Expression
+import pl.touk.nussknacker.restmodel.CustomActionRequest
+
+class CustomActionValidator(allowedAction: CustomActionDefinition) {
+
+  def validateCustomActionParams(
+      request: CustomActionRequest
+  ): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    implicit val nodeId: NodeId = NodeId(allowedAction.actionName.value)
+    val customActionParams      = allowedAction.parameters
+
+    val missingParams = checkForMissingParams(request.params, customActionParams)
+    val checkedParams = validateParams(request.params, customActionParams)
+
+    missingParams.combine(checkedParams)
+  }
+
+  private def checkForMissingParams(
+      requestParamsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    val paramsFromRequest = requestParamsMap.keys.toSet
+    val mandatoryParams = customActionParams.collect {
+      case param if param.validators.contains(MandatoryParameterValidator) && !paramsFromRequest.contains(param.name) =>
+        MandatoryParameterValidator.isValid(ParameterName(param.name), Expression.spel(""), None, None)
+    }
+    mandatoryParams match {
+      case Nil      => Validated.Valid(())
+      case nonEmpty => nonEmpty.traverse(_.toValidatedNel).map(_ => ())
+    }
+  }
+
+  private def validateParams(
+      requestParamsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    requestParamsMap match {
+      case paramsMap if paramsMap.nonEmpty =>
+        handleNonEmptyParamsRequest(paramsMap, customActionParams)
+      case emptyMap if emptyMap.isEmpty =>
+        handleEmptyParamsRequest(customActionParams)
+    }
+  }
+
+  private def handleNonEmptyParamsRequest(
+      paramsMap: Map[String, String],
+      customActionParams: List[CustomActionParameter]
+  )(implicit nodeId: NodeId): ValidatedNel[PartSubGraphCompilationError, Unit] = {
+
+    paramsMap.toList.map { case (name, expression) =>
+      customActionParams.find(_.name == name) match {
+        case Some(param) => toValidatedNel(param, expression, name)
+        case None =>
+          MismatchParameter(
+            s"Couldn't find a matching parameter in action definition for this param: $name",
+            "",
+            ParameterName(name),
+            nodeId.id
+          ).invalidNel[Unit]
+      }
+    }.sequence_
+  }
+
+  private def toValidatedNel(param: CustomActionParameter, expressionValue: String, parameterName: String)(
```

Comment:
This name is confusing. It looks like you do some dumb conversion here but in fact, here is done the core of the logic - validations

Reply:
Changed the name.